module/zfs/zfs_sysfs.c

```diff
@@ -0,0 +1,550 @@
+			(void) snprintf(number, sizeof (number), "%llu",
+			    (u_longlong_t)property->pd_numdefault);
+			show_str = number;
+			break;
+		case PROP_TYPE_STRING:
+			show_str = property->pd_strdefault;
+			break;
+		case PROP_TYPE_INDEX:
+			(void) zprop_index_to_string(property->pd_propnum,
+			    property->pd_numdefault, &show_str,
+			    property->pd_types);
+			break;
+		default:
+			return (0);
+		}
+	} else {
+		return (0);
+	}
+
+	return (snprintf(buf, buflen, "%s\n", show_str));
+}
+
+static ssize_t
+dataset_property_show(struct kobject *kobj, struct attribute *attr, char *buf)
+{
+	zfs_prop_t prop = zfs_name_to_prop(kobject_name(kobj));
+	zprop_desc_t *prop_tbl = zfs_prop_get_table();
+	ssize_t len;
+
+	ASSERT3U(prop, <, ZFS_NUM_PROPS);
+
+	len = zprop_sysfs_show(attr->name, &prop_tbl[prop], buf, PAGE_SIZE);
+
+	return (len);
+}
+
+static ssize_t
+pool_property_show(struct kobject *kobj, struct attribute *attr, char *buf)
+{
+	zpool_prop_t prop = zpool_name_to_prop(kobject_name(kobj));
+	zprop_desc_t *prop_tbl = zpool_prop_get_table();
+	ssize_t len;
+
+	ASSERT3U(prop, <, ZPOOL_NUM_PROPS);
+
+	len = zprop_sysfs_show(attr->name, &prop_tbl[prop], buf, PAGE_SIZE);
+
+	return (len);
+}
+
+/*
+ * ZFS kernel feature attributes for '/sys/module/zfs/features.kernel'
+ *
+ * This list is intended for kernel features that don't have a pool feature
+ * association or that extend existing user kernel interfaces.
+ *
+ * A user processes can easily check if the running zfs kernel module
+ * supports the new feature.
+ *
+ * For example, the initial channel_program feature was extended to support
+ * async calls (i.e. a sync flag). If this mechanism were in place at that
+ * time, we could have added a 'channel_program_async' to this list.
+ */
+static const char *zfs_features[]  = {
+	/* --> Add new kernel features here (post ZoL 0.8.0) */
+};
+
+#define	ZFS_FEATURE_COUNT	ARRAY_SIZE(zfs_features)
+
+static ssize_t
+kernel_feature_show(struct kobject *kobj, struct attribute *attr, char *buf)
+{
+	return (snprintf(buf, PAGE_SIZE, "supported\n"));
+}
+
+static int
+zfs_kernel_features_init(zfs_mod_kobj_t *zfs_kobj, struct kobject *parent)
+{
+	int err;
+
+	err = zfs_kobj_init(zfs_kobj, ZFS_FEATURE_COUNT, 0,
+	    kernel_feature_show);
+	if (err)
+		return (err);
+
+	for (int f = 0; f < ZFS_FEATURE_COUNT; f++)
+		zfs_kobj_add_attr(zfs_kobj, f, zfs_features[f]);
+
+	err = zfs_kobj_add(zfs_kobj, parent, ZFS_SYSFS_KERNEL_FEATURES);
+	if (err)
+		zfs_kobj_release(&zfs_kobj->zko_kobj);
+
+	return (err);
+}
+
+/*
+ * Show the content for the give zfs pool feature
+ */
+static ssize_t
+pool_feature_show(struct kobject *kobj, struct attribute *attr, char *buf)
+{
+	spa_feature_t fid;
+	ssize_t len = 0;
+
+	/* use the pool feature description for content */
+	if (zfeature_lookup_guid(attr->name, &fid) == 0 &&
+	    spa_feature_table[fid].fi_desc != NULL) {
+		len = snprintf(buf, PAGE_SIZE, "%s\n",
+		    spa_feature_table[fid].fi_desc);
+	}
```

Comment:
It would be nice if this was symmetric with properties with each feature being a subdirectory.  This would help future proof the directory layout and allow you to add attributes for `fi_flags` and `fi_depends` both of which I think would be useful additions.

Reply:
great idea. fixed in latest push